tests/test_cuda_bd_exchange_mover.py

```diff
@@ -0,0 +1,192 @@
+from timemachine.md.exchange.exchange_mover import BDExchangeMove as RefBDExchangeMove
+from timemachine.potentials import HarmonicBond, Nonbonded
+
+
+@pytest.mark.memcheck
+@pytest.mark.parametrize("moves", [1, 2, 10])
+@pytest.mark.parametrize("precision,rtol,atol", [(np.float64, 5e-7, 5e-7), (np.float32, 1e-6, 2e-6)])
+@pytest.mark.parametrize("seed", [2023])
+def test_two_clashy_water_moves(moves, precision, rtol, atol, seed):
+    """Given two waters directly on top of each other in a box, the exchange mover should accept almost any move"""
+    ff = Forcefield.load_default()
+    system, conf, _, _ = builders.build_water_system(1.0, ff.water_ff)
+    bps, _ = openmm_deserializer.deserialize_system(system, cutoff=1.2)
+
+    nb = next(bp for bp in bps if isinstance(bp.potential, Nonbonded))
+    bond_pot = next(bp for bp in bps if isinstance(bp.potential, HarmonicBond)).potential
+
+    all_group_idxs = get_group_indices(get_bond_list(bond_pot), conf.shape[0])
+
+    # Get first two mols
+    group_idxs = all_group_idxs[:2]
+
+    conf_idxs = np.array(group_idxs).reshape(-1)
+
+    conf = conf[conf_idxs]
+    # Set the two waters ontop of each other
+    conf[group_idxs[0], :] = conf[group_idxs[1], :]
+
+    box = np.eye(3) * 100.0
+
+    N = conf.shape[0]
+
+    params = nb.params[conf_idxs]
+
+    cutoff = nb.potential.cutoff
+    klass = custom_ops.BDExchangeMove_f32
+    if precision == np.float64:
+        klass = custom_ops.BDExchangeMove_f64
+
+    proposals_per_move = 1
+    bdem = klass(N, group_idxs, params, DEFAULT_TEMP, nb.potential.beta, cutoff, seed, proposals_per_move)
+
+    ref_bdem = RefBDExchangeMove(nb.potential.beta, cutoff, params, group_idxs, DEFAULT_TEMP)
+
+    assert bdem.last_log_probability() == 0.0, "First log probability expected to be zero"
+    last_conf = conf
+    accepted = 0
+    for _ in range(moves):
+        before_log_weights = ref_bdem.batch_log_weights(last_conf, box)
+        x_move, x_box = bdem.move(last_conf, box)
+        after_log_weights = ref_bdem.batch_log_weights(x_move, x_box)
+        # The box will never change
+        np.testing.assert_array_equal(box, x_box)
+        num_moved = 0
+        for mol_idxs in group_idxs:
+            if not np.all(x_move[mol_idxs] == last_conf[mol_idxs]):
+                num_moved += 1
+        if num_moved > 0 and proposals_per_move == 1:
+            accepted += 1
+            # Verify that the probabilities agree when we do accept moves
+            np.testing.assert_allclose(
+                np.exp(bdem.last_log_probability()),
+                np.exp(np.minimum(logsumexp(before_log_weights) - logsumexp(after_log_weights), 0.0)),
+                rtol=rtol,
+                atol=atol,
+            )
+        if num_moved == 0:
+            np.testing.assert_array_equal(last_conf, x_move)
+        assert num_moved <= 1, "More than one mol moved, something is wrong"
+        last_conf = x_move
+    assert accepted > 0
```

Comment:
What's the actual number of accepted moves? (The description says "Given two waters directly on top of each other in a box, the exchange mover should accept almost any move") This sounds like its testing for accepting at least one move?

Reply:
Made it stricter to accept all moves, though that is not strictly always true. So changing the seed could break tests https://github.com/proteneer/timemachine/pull/1202/commits/038c1564c5d607e2f871c54986cb0ecb559d7154